ENH: Sphere models and dipole fitting

Eric89GXL opened the pull request:
Finally getting around to adding some functionality that MNE-C has that we don't. Here's on the todo list:

1. ~~Improve head surface finding algorithm to check for both the BEM and the `-head.fif` file.~~

2. ~~Add code to create a spherical model for MEG (easy) and EEG (annoying).~~

3. ~~Add code to forward solution to properly deal with spherical models instead of BEMs.~~

4. ~~Add better dipole I/O functions.~~

5. ~~Add dipole fitting.~~

6. ~~Add `ad_hoc` noise~~

This PR's going to be big, sorry... but at least the coverage should be good.

Eric89GXL commented:
cc @staulu if you want to see how it's going

agramfort commented on doc/source/python_reference.rst:
```diff
@@ -139,7 +139,7 @@ Functions:
    read_bem_surfaces
    read_cov
-   read_dip
+   read_dipoles
```
I agree we should rename but we should change the API too. We should create a Dipole class with a plot method. That will fix our problem with @yousrabk on the RAP-MUSIC code which could then simply return a list of dipole instances with orientation information. 

cc @joewalter

Eric89GXL replied:
Fine by me, should be easy enough. It could just have a few attributes, a nice `__repr__`, and a `.plot()` method. I assume you're thinking a `Dipole` is a single dipole with orientation and magnitude as a function of time?

agramfort replied:
> Fine by me, should be easy enough. It could just have a few attributes, a nice __repr__, and a .plot() method. I assume you're thinking a Dipole is a single dipole with orientation and magnitude as a function of time?

yes exactly. It should be done in another PR

Eric89GXL replied:
Well I've already changed the API a bit to return a single `dict` instead of a tuple of pos, orientation, amplitude, etc. It will be very easy to make it a simple object with attributes, so I'll at least do that.

agramfort replied:
>
> Well I've already changed the API a bit to return a single dict instead of a tuple of pos, orientation, amplitude, etc. It will be very easy to make it a simple object with attributes, so I'll at least do that.
>
>
so Dipole class would simply inherit from dict with a repr + cplot method.

can you isolate the change in a new PR?

joewalter replied:
> We should create a Dipole class with a plot method
This is interesting for all sparse inverse methods. For the plot function, we should think of a way to plot not only the location but also the orientation per per dipole and time point similar to the way of selecting a time point in our visualization of dense solvers

Eric89GXL replied:
I plan on changing a bit about how dipoles are processed, but I could at least isolate the new reading and writing functions if that would make your life easier, yeah

Eric89GXL commented on mne/bem.py:
```diff
@@ -0,0 +1,300 @@
+
+@verbose
+def fit_sphere_to_headshape(info, dig_kinds=(FIFF.FIFFV_POINT_EXTRA,),
+                            verbose=None):
```
These two functions are just moved

Eric89GXL commented on mne/dipole.py:
```diff
@@ -43,4 +83,33 @@ def read_dip(fname, verbose=None):
     gof = data[:, 9]
-    return time, pos, amplitude, ori, gof
+    dipoles = dict(time=time, pos=pos, amplitude=amplitude, ori=ori,
+                   gof=gof, name=name)
```
@agramfort this is what I changed `read_dipoles` to return (and `read_dip` splits it back out into a tuple), so I can just make it a class

agramfort replied:
yeah. I can have a look this afternoon if you want. It should be quick.

Eric89GXL replied:
I am going to bed now and I probably won't get around to creating the class until next week sometime, so feel free to open a PR into my branch if you want

Eric89GXL commented on mne/preprocessing/maxfilter.py:
```diff
@@ -8,78 +8,11 @@
-
-@verbose
-def fit_sphere_to_headshape(info, dig_kinds=(FIFF.FIFFV_POINT_EXTRA,),
-                            verbose=None):
```
moved, this wasn't a great place for it anyway

agramfort commented:
let us know when it's ready to review.

Eric89GXL commented:
The optimization takes forever. It needs to compute a bunch of single-vertex forward solutions. In C it isn't a problem, but for us it is. Our Python forward solution code is fast(er than C) only because of how it's vectorized -- the non-vectorized versions were brutally slower than C. And computing one vertex forward, then another, then another... is partially un-vectorizing it. It takes about a minute to fit one time point on my machine compared to less than a second in C. Any ideas?

One idea is, instead of allowing for arbitrary locations, compute the forward solution on some suitably small predefined grid, then use linear interpolation between the pre-calculated forward solutions to get forward solutions for arbitrary positions -- or just restrict the possible dipole locations to that original grid. I don't love the idea, but it might be the only way to make it reasonable in Python without moving to Cython.

agramfort commented:
hum... leadfield interpolation is reasonable and used in various software.

can we parallelize over time points?

Eric89GXL commented:
Yeah, but it's still terrible even with that :)

agramfort commented:
Caching?

> On 27 févr. 2015, at 23:17, Eric Larson <notifications@github.com> wrote:
> 
> Yeah, but it's still terrible even with that :)
> 
> —
> Reply to this email directly or view it on GitHub.
>

Eric89GXL commented:
Not sure how you would cache, since there are (effectively) an infinite number of positions that could be used. Maybe field interpolation will be good enough. Maybe I'll start with that and see how tightly spaced the grid needs to be to get reasonable agreement with the C code.

adykstra commented:
What about an iterative procedure with progressively finer spacings?
On Feb 27, 2015 11:23 PM, "Eric Larson" <notifications@github.com> wrote:

> Not sure how you would cache, since there are (effectively) an infinite
> number of positions that could be used. Maybe field interpolation will be
> good enough. Maybe I'll start with that and see how tightly spaced the grid
> needs to be to get reasonable agreement with the C code.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/1828#issuecomment-76483903>.
>

Eric89GXL commented:
Hadn't thought of that, it could potentially work...

What the heck, maybe I'll try both interpolation and iterative refinement (or whatever such an algorithm would be called) and compare them to each other and the C code.

agramfort commented:
fun we in perspective :)

adykstra commented:
@Eric89GXL, will this PR apply to EEG as well as MEG data?  Also, multi dipole fitting?

Eric89GXL commented:
EEG and MEG, yeah, but only single dipoles at first

jona-sassenhagen commented:
Fantastic! Skimming over this, it seems it takes evoked data. How do you feel about allowing it to also localise independent components? It seems it would only require changing a few lines in fit_dipole.

Eric89GXL commented:
That would probably be a trivial change once I get it all working, but I'll definitely leave it to a subsequent PR :)

jona-sassenhagen commented:
Okay, I can try it myself once this one is merged.

Eric89GXL commented:
@agramfort this is ready for a first round of review.

I plan on tweaking the code a tiny bit more to see if I can get the goodness of fits to be better. For some reason I can get better fits (in terms of absolute position differences and orientation differences) in mne-python, but with lower goodness of fit estimates than the C code in some cases. Not sure why... but in any case, it seems like in most conditions we get the same or better performance in Python compared to C, at least with simulated data.

agramfort commented:
I'll look this we if I can.

Do you support noise cov?

how are we in terms of speed?

Eric89GXL commented:
Slower than C, but it's not so bad. Fit about 1 dipole per second (python) instead of 5 or 10 (C) for the BEM case. Spherical is faster, naturally. The slowness comes from the forward solution calculations, so there isn't much we can do about it.

Yep, it supports noise covariance / whitening, just like the C code. I guess the one thing it doesn't have is "ad hoc" noise as Matti calls it, but we can almost trivially add a function to create that type of noise cov. Maybe I should do that, actually... In any case, I started out replicating the C code, but switched it over to use our whitening, which is slightly different numerically, but gives about the same results.

agramfort commented:
cool !

I'll have a look asap

Eric89GXL commented on .travis.yml:
```diff
@@ -9,9 +9,9 @@ env:
+    - PYTHON=2.7 DEPS=full TEST_LOCATION=src
     - PYTHON=2.7 DEPS=full TEST_LOCATION=install
-    - PYTHON=2.7 DEPS=nodata TEST_LOCATION=src  # also runs flake8
+    - PYTHON=2.7 DEPS=nodata TEST_LOCATION=src MNE_DONTWRITE_HOME=true  # also runs flake8
```
Also FYI this fixes our coverage. Turns out the 2.7 full deps src version wasn't actually running any of the tests that required testing data, so after this PR our coverage should jump back up to 90+.

agramfort commented on mne/bem.py:
```diff
@@ -0,0 +1,321 @@
+    for k in range(u['nterms'] - 1):
+        k1 = k + 1
+        mu1n = np.power(mu[0], k1)
+        u['y'][k] = u['w'][k] * (u['fn'][k1] - mu1n*u['fn'][0])
```
pep8

Eric89GXL replied:
Travis automatically checks PEP8 now, so there really shouldn't be an error, what did you have in mind?

Eric89GXL replied:
ohhh the operators, forgot about that. FYI scipy has stopped following that convention and it is disabled by default in many PEP8 checkers. If you want to keep it, I'll fix it and enable it in Travis.

agramfort commented on mne/bem.py:
```diff
@@ -0,0 +1,321 @@
+                         'jac': lambda x: np.array([0.] * ii + [val] +
+                                                   [0.] * (nfit - ii - 1))})
+    mu = optimize.minimize(fun, mu_0, constraints=cons, method='COBYLA',
+                           tol=1e-2).x
```
did you play with other solvers?

Eric89GXL replied:
Not too extensively here. This is just fitting the EEG sphere model, and seems to work fine. I did choose this before extensively testing the dipole-fitting solver, and the best one there ended up being COBYLA so perhaps this isn't a terrible choice.

agramfort commented on mne/bem.py:
```diff
@@ -0,0 +1,321 @@
+                  coord_frame=FIFF.FIFFV_COORD_HEAD)
+    sphere['layers'] = []
+    if head_radius is not None:
+        # Eventually these could be configurable...
```
+1. Let's expose it as parameters

Eric89GXL replied:
The radii and sigmas? Are these often changed by people? Note that the radii here are relative to the actual radius, which is already a parameter the user can change.

agramfort commented on mne/bem.py:
```diff
@@ -0,0 +1,321 @@
+    Returns
+    -------
+    bem : dict
+        A spherical BEM.
```
I don't like the BEM naming for an analytical sphere solution

is it called BEM is the C code?

Eric89GXL replied:
No, I'll change it to "sphere", it's what I used every time I created one anyway as `sphere = make_sphere_model(..)`

agramfort commented on mne/bem.py:
```diff
@@ -0,0 +1,321 @@
+    -------
+    radius : float
+        Sphere radius in mm.
+    origin_head: ndarray
```
shape?

Eric89GXL replied:
this code was copy-pasted, but I can add this info

agramfort commented on mne/bem.py:
```diff
@@ -0,0 +1,321 @@
+        Sphere radius in mm.
+    origin_head: ndarray
+        Head center in head coordinates (mm).
+    origin_device: ndarray
```
shape?

agramfort commented on mne/bem.py:
```diff
@@ -0,0 +1,321 @@
+    xradius = (np.max(points[:, 0]) - np.min(points[:, 0])) / 2.
+    yradius = (np.max(points[:, 1]) - np.min(points[:, 1])) / 2.
+
+    radius_init = (xradius + yradius) / 2
```
2.

agramfort commented on mne/dipole.py:
```diff
@@ -138,11 +173,285 @@ def read_dipole(fname, verbose=None):
+    Returns
+    -------
+    dip : instance of Dipole
+        The dipole fits.
```
I'd love to have an option to return the residual like we do it in the sparse solvers

agramfort replied:
and it would allow to fit dipoles iteratively

Eric89GXL replied:
That should be a fairly simple subsequent PR :)

Eric89GXL commented on mne/forward/_make_forward.py:
```diff
@@ -177,19 +291,19 @@ def make_forward_solution(info, mri, src, bem, fname=None, meg=True, eeg=True,
+    bem : dict | str
         Filename of the BEM (e.g., "sample-5120-5120-5120-bem-sol.fif") to
-        use.
+        use, or a loaded BEM (dict).
```
reminder for me to change this to `a loaded sphere model (dict)`

Eric89GXL commented:
Comments addressed. I added the residuals as an output. I also moved a few lines so that it's easy to see how it could be applied to other bits of data, e.g. raw. For future compatibility, do we maybe want a `fit_dipole` function that takes `data, info, times` instead of `evoked`, and then have `fit_dipole_evoked` that is just a simple wrapper?

@agramfort the only change I didn't implement was exposing the relative radii and sigmas of the sphere model as parameters. I'd rather leave it until someone asks for it -- it might be soon, but I haven't really messed with the parameters enough to feel comfortable exposing the functionality.

Eric89GXL commented:
@agramfort feel free to review one final time / merge. This touches a bunch of files so I don't want it open too long if possible.

agramfort commented:
will do as soon as possible

agramfort commented:
we'll have a _fit_dipole that takes data, info, times in the future.

We'll add a fit_dipole_raw to epilepsy data.

Ok for radii and sigmas

agramfort commented on examples/inverse/plot_dipole_comparison.py:
```diff
@@ -0,0 +1,136 @@
+src_nn = fwd['source_nn']
+
+# MNE-C skips the last "time" point :(
+dip.crop(dip_c.times[0], dip_c.times[-1])
```
AttributeError: 'tuple' object has no attribute 'crop'

agramfort replied:
also this example is good to debug but not to explain how to do dipole fitting to new users.

We should merge the 2 dipole examples and show how to do dipole fitting in 15 lines of code

We can keep this example in a gist once we're happy

agramfort replied:
i'll get back to review this when you fix the examples

there is no reason to advertise mne_dipole_fit anymore in the other example.

Eric89GXL replied:
IIRC my motivation behind doing this was first because the dipole fits for the real data were not very good (low GOF values). Then I decided I might as well make it a comparison. Is it worth showing on the sample dataset even if the GOF values are not good?

Eric89GXL replied:
Or can you think of a good dataset I should use instead?

agramfort replied:
if you use only the right channels on sample event_id=1 you should have a great goodness of fit for N100

crop the evoked around N100 to get high values.

do you agree that this example should not be merged?

Eric89GXL replied:
Yeah, this example is more of a sanity check for us I suppose. :)

I'll make a `gist` and link to it instead.

Eric89GXL commented:
@agramfort example simplified. I had to manually baseline correct the evoked object. Should I write a little method for `Evoked` objects like `evoked.baseline(tmin, tmax)`?

Eric89GXL commented:
Fixed the issue with the baselining by doing it in `read_evokeds`.

agramfort commented:
when running example I get

```
/Users/alex/local/lib/python/site-packages/mne/dipole.py in fit_dipole(evoked, cov, bem, trans, n_jobs, verbose)

/Users/alex/local/lib/python/site-packages/mne/utils.py in verbose(function, *args, **kwargs)
    535         finally:
    536             set_log_level(old_level)
--> 537     return function(*args, **kwargs)
    538
    539

/Users/alex/local/lib/python/site-packages/mne/dipole.py in fit_dipole(evoked, cov, bem, trans, n_jobs, verbose)
    383         mri_head_t = {'from': FIFF.FIFFV_COORD_HEAD,
    384                       'to': FIFF.FIFFV_COORD_MRI, 'trans': np.eye(4)}
--> 385     bem = _setup_bem(bem, bem, neeg, mri_head_t)
    386     if not bem['is_sphere']:
    387         if trans is None:

/Users/alex/local/lib/python/site-packages/mne/forward/_make_forward.py in _setup_bem(bem, bem_extra, neeg, mri_head_t)
    181     else:
    182         if neeg > 0 and len(bem['surfs']) == 1:
--> 183             raise RuntimeError('Cannot use a homogeneous model in EEG '
    184                                'calculations')
    185         logger.info('Employing the head->MRI coordinate transform with the '

RuntimeError: Cannot use a homogeneous model in EEG calculations
```

Eric89GXL commented:
Sorry, that was a silly change I made while testing, just needed to have `eeg=False` in the picking. Try now.

agramfort commented:
works now thanks !

agramfort commented:
Ran 2 tests in 21.165s

on test_dipole.py

any chance to make it faster? using less channels? less time points?

also update what's new

than +1 for merge.

what you did @Eric89GXL is huge !

Eric89GXL commented:
Yikes, I think on my system it was about 6 seconds. I'll take a look at making it faster tomorrow.

agramfort commented:
looks like you need to rebase.

Eric89GXL commented:
On my system it's actually 4 seconds, which seems reasonable because it's calculating forward solutions... is it really 21 on yours? How long does `mne/forward/tests/test_make_forward.py` take?

Eric89GXL commented:
It even only takes 11 seconds on Travis, which is usually slower than desktops (!)

Eric89GXL commented:
I cut the number of channels by a factor of 3, tests now take 2.1 seconds on my machine. Hopefully it's better for you, too.

agramfort commented on doc/source/whats_new.rst:
```diff
@@ -9,6 +9,10 @@ Current
 
+   - Add support for single dipole fitting by `Eric Larson`_
+
+   - Add support for spherical BEM models by `Eric Larson`_
```
again spherical and BEM in one sentence does not make sense to me. BEM refers to boundary element method and numerical integration. I would just say

Add support for spherical forward models by `Eric Larson`_

Eric89GXL replied:
It's still a boundary element model (what matters is the changes at the boundary borders), it just uses a spherical surface. Note e.g. Matti has a function called `mne_make_sphere_bem`...

Eric89GXL replied:
... and the fact that you know the surfaces are spherical simplifies some of the math. But I think it still qualifies as a BEM, unless I have my definitions mixed up

agramfort replied:
do you have triangles on a sphere mesh?

or you use sarvas formula for MEG?

agramfort commented:
besides lgtm

please rebase before merge

Eric89GXL commented:
I already rebased, so I'll just change the spherical wording to be more conservative, and merge once I see green.

agramfort commented:
merge if you're done @Eric89GXL

Eric89GXL commented:
Just waiting to see green

agramfort commented:
:beers: !

> On 12 mars 2015, at 19:06, Eric Larson <notifications@github.com> wrote:
> 
> Merged #1828.
> 
> —
> Reply to this email directly or view it on GitHub.
>
